src/DynamoCore/DSEngine/NodeToCode.cs

```diff
@@ -1,16 +1,692 @@
+                }
+                else if (astNode is FunctionCallNode)
+                {
+                    var node = astNode as FunctionCallNode;
+                    for (int i = 0; i < node.FormalArguments.Count; ++i)
+                    {
+                        Visit(node.FormalArguments[i], func);
+                    }
+                    Visit(node.ArrayDimensions, func);
+                }
+                else if (astNode is ArrayNode)
+                {
+                    var node = astNode as ArrayNode;
+                    Visit(node.Expr, func);
+                    Visit(node.Type, func);
+                }
+                else if (astNode is ExprListNode)
+                {
+                    var node = astNode as ExprListNode;
+                    for (int i = 0; i < node.list.Count; ++i)
+                    {
+                        Visit(node.list[i], func);
+                    }
+                    Visit(node.ArrayDimensions, func);
+                }
+                else if (astNode is FunctionDotCallNode)
+                {
+                    var node = astNode as FunctionDotCallNode;
+                }
+                else if (astNode is InlineConditionalNode)
+                {
+                    var node = astNode as InlineConditionalNode;
+                    Visit(node.ConditionExpression, func);
+                    Visit(node.TrueExpression, func);
+                    Visit(node.FalseExpression, func);
+                }
+                else if (astNode is RangeExprNode)
+                {
+                    var node = astNode as RangeExprNode;
+                    Visit(node.FromNode, func);
+                    Visit(node.ToNode, func);
+                    Visit(node.StepNode, func);
+                    Visit(node.ArrayDimensions, func);
+                }
+                else if (astNode is BinaryExpressionNode)
+                {
+                    var node = astNode as BinaryExpressionNode;
+                    Visit(node.LeftNode, func);
+                    Visit(node.RightNode, func);
+                }
+            }
+        }
+
+        /// <summary>
+        /// The numbering state of a variable.
+        /// </summary>
+        private class NumberingState
+        {
+            /// <summary>
+            /// Variable name.
+            /// </summary>
+            public string Variable { get;  private set; }
+
+            /// <summary>
+            /// Variable name with number suffix.
+            /// </summary>
+            public string NumberedVariable
+            {
+                get { return Variable + ((ID == 0) ? String.Empty : ID.ToString()); }
+            }
+
+            /// <summary>
+            /// Number suffix.
+            /// </summary>
+            public int ID { get; set; }
+
+            /// <summary>
+            /// Indicate it is for ASTs in a new code block
+            /// </summary>
+            public bool IsNewSession { get; set; }
```

Comment:
The logic here is a little bit complicated, I'll explain it later. But the class `NumberState` is a private class of `NodeToCodeUtils'. Besides, the node to code logic should be cleaned up later on, so no worry, i'll update it later :-)